### ISSUE TITLE
Memory usage appears to increase if dtype is cast after Tensor.expand

### ISSUE THREAD

**Issue by Horse7354** [pytorch/pytorch#116142]
### 🐛 Describe the bug

Hi, I was debugging a memory issue in a custom module, and encountered the following strange memory behavior; Changing dtype from fp16 to fp32 shows an increase in memory usage on a tensor that was previously expanded in shape.
I am not sure if this is indeed a bug or if I am misunderstanding, but this seems to have practical implications where I get OOM in one case and not the other.

Code illustrating issue:

```python
import torch

x = torch.randn(100, dtype=torch.float16, device="cuda")
y = torch.randn(100, dtype=torch.float16, device="cuda")

# memory is the same before and after dtype change
print("before x fp32\n", torch.cuda.memory_summary())
x = x.to(torch.float32)
print("after x fp32\n", torch.cuda.memory_summary())

x = x.unsqueeze(-1).expand(-1, 10000)

y = y.unsqueeze(-1).expand(-1, 10000)

# memory before the dtype change is the same as the previous 2 print statements, but memory after the dtype change increases substantially
print("before y fp32\n", torch.cuda.memory_summary())
y = y.to(torch.float32)
print("after y fp32\n", torch.cuda.memory_summary())
```

### Versions

Versions of relevant libraries:
```
[pip3] flake8==4.0.1
[pip3] flake8-polyfill==1.0.2
[pip3] gpytorch==1.11
[pip3] mypy==0.960
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.22.4
[pip3] torch==2.0.1
[pip3] torchvision==0.15.2
[pip3] triton==2.0.0
[conda] Could not collect
```

**Comment by vadimkantorov** [comment]
The second dtype change probably will lead to reallocation/materialization of the full expanded tensor and will make it contiguous. I guess this is the reason

**Comment by Horse7354** [comment]
> The second dtype change probably will lead to reallocation/materialization of the full expanded tensor and will make it contiguous. I guess this is the reason

This makes sense, however would that not make it a bug?

A similar thing happens with nn.functional.softmax() after an expand, and unfortunately unlike simply changing when I do the .to(), the extra expanded dimensions are required to have in the softmax for my use case. Here is example code for similar memory issues with softmax:

```python
import torch

x = torch.randn(100, dtype=torch.float16, device="cuda")
y = torch.randn(100, dtype=torch.float16, device="cuda")

print("before fp32\n", torch.cuda.memory_summary())
x = x.to(torch.float32)
y = y.to(torch.float32)
print("after fp32\n", torch.cuda.memory_summary()) # small memory increase
x = torch.nn.functional.softmax(x, dim=0)
print("after x softmax\n", torch.cuda.memory_summary()) # no memory increase

x = x.unsqueeze(-1).expand(-1, 10000)

y = y.unsqueeze(-1).expand(-1, 10000)

print("before y softmax\n", torch.cuda.memory_summary()) # same memory as last print statement
y = torch.nn.functional.softmax(y, dim=0)
print("after y softmax\n", torch.cuda.memory_summary()) # large memory increase
```

**Comment by zou3519** [comment]
This is expected due to @vadimkantorov's comment.

**Comment by Horse7354** [comment]
For future reference I was able to get around this behavior in practice by not using `Tensor.expand`, and by instead using `torch.einsum` for all the matrix ops that would have required unsqueezing or expanding dims. 
Maybe the memory phenomenon at least merits a mention or warning in the docs for `Tensor.expand`, that downstream ops (like even casting dtype) may reallocate the redundant memory.

**Comment by vadimkantorov** [comment]
For materialization behavior of `.to()`, changing it would be a breaking change and changing it could uncover some bugs / reliance on this in user code, but I guess this semantics would be best discussed in a separate issue

**Comment by janeyx99** [comment]
Documentation seems to be the clearest next step, though it would be interesting to figure out where to best add docs. We can start by adding a warning to `tensor.expand` as @Horse7354 recommended.

This fact seems to be general enough to live in a more overview place about memory, though I do not know if we already have existing docs about strides, overlapping memory, and other memory information.
